artemis-core-client/src/main/java/org/apache/activemq/artemis/core/client/impl/ClientProducerImpl.java

```diff
@@ -410,108 +410,117 @@ private void largeMessageSendBuffered(final boolean sendBlocking,
                   sendInitialLargeMessageHeader(msgI, credits);
                }
-               int creditsSent = sessionContext.sendLargeMessageChunk(msgI, messageSize.get(), sendBlocking, true, buff, reconnectID, handler);
+               final long messageBodySize = deflaterReader != null ? deflaterReader.getTotalSize() : totalBytesRead;
+               int creditsSent = sessionContext.sendLargeMessageChunk(msgI, messageBodySize, sendBlocking, true, trimmedBuffer(scratchBuffer, bytesRead), reconnectID, handler);
                credits.acquireCredits(creditsSent);
+               break;
             }
-         } else {
+            // !lastPacket
             if (!headerSent) {
                headerSent = true;
                sendInitialLargeMessageHeader(msgI, credits);
             }
 
-            int creditsSent = sessionContext.sendLargeMessageChunk(msgI, messageSize.get(), sendBlocking, false, buff, reconnectID, handler);
+            int creditsSent = sessionContext.sendLargeMessageChunk(msgI, 0, sendBlocking, false, trimmedBuffer(scratchBuffer, bytesRead), reconnectID, handler);
             credits.acquireCredits(creditsSent);
          }
-      }
-
-      try {
-         input.close();
       } catch (IOException e) {
          throw ActiveMQClientMessageBundle.BUNDLE.errorClosingLargeMessage(e);
       }
    }
+
+   /**
+    * This is trimming {@code buffer} to the expected size or reusing it if not needed.
+    */
+   private static byte[] trimmedBuffer(byte[] buffer, int expectedSize) {
+      if (buffer.length == expectedSize) {
+         return buffer;
+      }
+      byte[] trimmedBuffer = new byte[expectedSize];
+      System.arraycopy(buffer, 0, trimmedBuffer, 0, expectedSize);
+      return trimmedBuffer;
+   }
+
+   private static boolean isLargeMessageLastPacket(int readResult) {
+      return readResult <= 0;
+   }
+
+   private static int bytesRead(int readResult) {
+      return readResult > 0 ? readResult : -readResult;
+   }
+
+   /**
+    * Use {@link #isLargeMessageLastPacket(int)} and {@link #bytesRead(int)} to decode the result of this method.
+    */
+   private static int readLargeMessageChunk(InputStream inputStream,
+                                            byte[] readBuffer,
+                                            int chunkLimit) throws ActiveMQLargeMessageException {
+      assert chunkLimit > 0;
+      int bytesRead = 0;
+      do {
+         final int remaining = chunkLimit - bytesRead;
+         final int read;
+         try {
+            read = inputStream.read(readBuffer, bytesRead, remaining);
+         } catch (IOException e) {
+            throw ActiveMQClientMessageBundle.BUNDLE.errorReadingBody(e);
+         }
+         if (read == -1) {
+            // bytesRead can be 0 if the stream return -1 after 0-length reads
+            return -bytesRead;
+         }
+         bytesRead += read;
+      }
+      while (bytesRead < chunkLimit);
+      assert bytesRead == chunkLimit;
```

Comment:
is this assert correct? what if you are reading a large message that Is not a multiple of chunkLimit?